File: examples/dangling_pointer_closure.cpp
#include <iostream>
#include <functional>

using namespace std;

function<int(int)> get_lambda_with_local_reference() {
    int local_arr[] = { 1, 2, 3, 4, 5 };
    return [&](int value) { return value + local_arr[2]; };
}

int main() {
    // This function returns a lambda which internally keeps a pointer to a local array.
    // Of course, when using the returned lambda, the array does not exist anymore.
    function<int(int)> function = get_lambda_with_local_reference();
    cout << "lambda uses stack-local reference:" << function(6) << endl;
}

File: examples/dangling_pointer_closure.rs
// This function returns a closure with a pointer to a stack-local array.
// The Rust Complainer says no.
fn get_lambda_with_local_reference() -> impl Fn(i32) -> i32 {
    let local_arr = [1, 2, 3];
    |value| value + local_arr[2]
}

// Moving the stack-local array into the returned closure is fine.
fn get_lambda_with_moved_reference() -> impl Fn(i32) -> i32 {
    let local_arr = [1, 2, 3];
    move |value| value + local_arr[2]
}

fn main() {
    let function = get_lambda_with_local_reference();
    println!("lambda uses stack-local reference: {}", function(6));
}
